src/python/foglamp/device/server.py

```diff
@@ -87,45 +92,50 @@ class Server:
         error = None
 
         try:
+            # Create Category before loading the plugin
             config = {}
             storage = Storage(cls._core_management_host, cls._core_management_port, svc=None)
             cfg_manager = ConfigurationManager(storage)
-            await cfg_manager.create_category(cls._plugin_name, config,
-                                                        '{} Device'.format(cls._plugin_name), True)
-
+            await cfg_manager.create_category(cls._plugin_name, config, '{} Device'.format(cls._plugin_name), True)
             config = await cfg_manager.get_category_all_items(cls._plugin_name)
 
             try:
                 plugin_module_name = config['plugin']['value']
             except KeyError:
-                _LOGGER.warning("Unable to obtain configuration of module for plugin {}".format(cls._plugin_name))
-                raise
+                _LOGGER.exception("Unable to obtain configuration of module for plugin {}".format(cls._plugin_name))
+                # Retrieve the plugin name and hence plugin module name, from the params
+                plugin_module_name = cls._plugin_name
 
             try:
                 cls._plugin = __import__("foglamp.device.{}_device".format(plugin_module_name), fromlist=[''])
             except Exception:
-                error = 'Unable to load module {} for device plugin {}'.format(plugin_module_name,
-                                                                               cls._plugin_name)
+                error = 'Unable to load module {} for device plugin {}'.format(plugin_module_name, cls._plugin_name)
                 raise
-
             default_config = cls._plugin.plugin_info()['config']
 
+            # Complete the Component definition with config info from the plugin
             await cfg_manager.create_category(cls._plugin_name, default_config,
-                                                        '{} Device'.format(cls._plugin_name))
+                                                        '{} Device'.format(cls._plugin_name, ), True)
 
             config = await cfg_manager.get_category_all_items(cls._plugin_name)
 
             # TODO: Register for config changes
 
             cls._plugin_data = cls._plugin.plugin_init(config)
+
+            if isinstance(cls._plugin_data, tuple):
+                # If plugin is an aiohttp listener, then return tuple (coro, handler) from plugin_init
+                coro, cls._plugin_server_handler = cls._plugin_data
+                if coro is not None:
+                    cls._plugin_server = await coro
+
```

Comment:
instead of this, should not we pass current event_loop as an arg? and let http listener plugin to add aiohttp start corutine added to it; 

Understood that device >server> start def run loop forever, but here adding it as coro should not be an issue; as it happens before!